Sources/MIDIKit/MIDINetwork.swift

```diff
@@ -0,0 +1,199 @@
+//
+//  MIDINetwork.swift
+//  
+//
+//  Created by Michele Longhi on 28/10/2020.
+//
+
+import Foundation
+import CoreMIDI
+
+extension MIDIClient {
+    
+    public convenience init(name: String, networkConnection connection: MIDINetworkConnection) {
+        self.init(name: name)
+        
+        startWithNetworkConnection(connection)
+    }
+    
+    public func startWithNetworkConnection(_ connection: MIDINetworkConnection) {
+        
+        removeAllNetworkConnections()
+        
+        networkSession = MIDINetworkSession()
+        networkSession?.addConnection(connection)
+        
+        lastKnownConnection = connection
+        
+        try? start()
+    }
+
+    public func startWithFirstAvailableNetworkConnection(useLastKnownConnection: Bool = true, completion handler: ((_ connection: MIDINetworkConnection?)->Void)? = nil) {
+        
+        removeAllNetworkConnections()
+        
+        let networkSession = self.networkSession ?? MIDINetworkSession()
+        
+        if let connection = lastKnownConnection, useLastKnownConnection {
+            networkSession.addConnection(connection)
+            try? start()
+
+            handler?(connection)
+            return
+        }
+    
+        bonjourService.findService(BonjourService.Services.Apple_Midi, domain: BonjourService.LocalDomain) { services in
```

Comment:
`bonjourService.findService(:domain:found:)` does not guarantee the Thread/RunLoop/DispatchQueue which it calls its callback `found`.
This alone results in a race condition.
In addition, `MIDIClient.start()` internally calls `MIDIClientCreateWithBlock` which needs to be called from a Thread that has a RunLoop which is not guaranteed by `bonjourService.findService(:domain:found:)`.
`MIDIClient.start()` currently misses this critical part of documentation, which is my fault.
Nether-less, it is really important because MDIClient will not receive Notifications if not started from a thread with a run loop. I just run into this issue a couple months ago.

Reply:
To be quite honest I’m not sure the service discovery has to be part of the MIDI client. Better to leave this task to the developer, using the preferred Bonjour implementation. I’ll scrap it from MIDIKit.